torch/_inductor/codecache.py

```diff
@@ -379,6 +388,163 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
     tmp_path.rename(path)
 
 
+def _reduce_fake_tensor(t):
+    """
+    See FxGraphCachePickler. Custom reducer to enable pickling FakeTensor.
+    """
+    return (TensorMetadataHolder, (_extract_tensor_metadata(t), t.device))
+
+
+@dataclasses.dataclass
+class TensorMetadataHolder:
+    tensor_metadata: TensorMetadata
+    device: torch.device
+
+
+def _reduce_typed_storage(s):
+    """
+    See FxGraphCachePickler. Custom reducer to enable pickling TypedStorage.
+    """
+    # The string repr seems to capture the relevant details for hashing, e.g.,
+    # 0.0\n[torch.storage.TypedStorage(dtype=torch.float32, device=cpu) of size 1]
+    return (TypedStorageMetadataHolder, (str(s),))
+
+
+@dataclasses.dataclass
+class TypedStorageMetadataHolder:
+    storage_metadata: str
+
+
+class FxGraphCachePickler(pickle.Pickler):
+    """
+    Custom pickler to customize the pickling of some objects, only for the purpose
+    of computing a hash for keying into the FxGraphCache. Some objects don't pickle
+    and/or contain attributes that vary between runs, and we want to capture the
+    data that allow us to compute a safe and stable hash.
+    """
+
+    dispatch_table = copyreg.dispatch_table.copy()
+
+    # Without custom pickling, we get:
+    # AttributeError: Can't pickle local object 'WeakValueDictionary.__init__.<locals>.remove'
+    dispatch_table[torch._subclasses.fake_tensor.FakeTensor] = _reduce_fake_tensor
+
+    # The _cdata attribute is not stable across runs.
+    dispatch_table[torch.storage.TypedStorage] = _reduce_typed_storage
+
+    @staticmethod
+    def dumps(obj) -> bytes:
+        """
+        Pickle an object using the FxGraphCachePickler.
+        """
+        with io.BytesIO() as stream:
+            pickler = FxGraphCachePickler(stream)
+            pickler.dump(obj)
+            return stream.getvalue()
+
+
+@dataclasses.dataclass
+class OrderedSetHolder:
+    items: List[Any]
+
+
+class FxGraphHashDetails:
```

Comment:
@eellison, note that I haven't included the guards in the data that I pickle/hash. I was getting: `'cannot pickle 'weakref.ReferenceType' object'`. I'm sure I can find a fix for that, but I thought it might be worth discussing whether it's necessary? I naively understand guards to capture Tensor shapes and the implementation here captures a lot of info about the tensors. For example:

`TensorMetadataHolder(tensor_metadata=TensorMetadata(shape=torch.Size([128, 64, 1, 1]), dtype=torch.float32, requires_grad=True, stride=(64, 1, 1, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={}), device=device(type='cuda', index=0))`

Is this enough? Or is my understanding too shallow and there are other kinds of guards?

Reply:
I don't think we need to include guards in the cache key. I think the guards only matter insofar as they affect what the graph looks like. 

I think for symbolic shapes it's a little more subtle, but I think the key we're caching on shouldn't be the guards but the symbolic shape env itself.

Reply:
I think for symbolic shapes the key should be the symbolic shape guards. They represent when the inductor graph is valid for reuse.

Reply:
Ok, I'm not grokking what I need to capture to make sure symbolic shapes are handled correctly. Forgive me if I'm way off, but I do notice that the Tensor metadata we capture in the existing implementation can look like this (if it's at all relevant):

`TensorMetadata(shape=torch.Size([s0, s1]), dtype=torch.bfloat16, requires_grad=False, stride=(s1, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={})`

Reply:
A proposal for a symbolic shape scheme:
- we should hash to make sure the same symbolic shape relationships appear in the same way e.g. `[s0, s1]` is different than `[s0, s0]`
- Serialize the guards we've added for each symbolic shape. 
- After you have a cache hit, at runtime test that the guards we've serialized pass for the runtime symbolic shapes

Reply:
See my TODO below. I'm wondering if it's maybe possibly sufficient to just consider shape_env.var_to_range

Reply:
synced with @eellison; the plan is to change this PR to handle static only and then do this proposal in a next step